### ISSUE TITLE
Creating variants with multiple option values does not work in Spree 3.0.0

### ISSUE THREAD

**Issue by vmatekole** [spree/spree#6331]
Creating variants with option values via API in Spree 3.0.0 seemingly does not work. 

Product JSON:
``` json
{
  "count": 1,
  "total_count": 1,
  "current_page": 1,
  "per_page": 25,
  "pages": 1,
  "products": [
    {
      "id": 225,
      "name": "Pineapple lamps",
      "description": null,
      "price": "100.0",
      "display_price": "£100.00",
      "available_on": "2015-04-25T13:11:36.178Z",
      "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
      "meta_description": null,
      "meta_keywords": null,
      "shipping_category_id": 1,
      "taxon_ids": [ ],
      "total_on_hand": 0,
      "has_variants": true,
      "master": {
        "id": 227,
        "name": "Pineapple lamps",
        "sku": "",
        "price": "100.0",
        "weight": "0.0",
        "height": null,
        "width": null,
        "depth": null,
        "is_master": true,
        "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
        "description": null,
        "track_inventory": true,
        "cost_price": null,
        "display_price": "£100.00",
        "options_text": "",
        "in_stock": false,
        "is_backorderable": true,
        "total_on_hand": 0,
        "is_destroyed": false,
        "option_values": [ ],
        "images": [ ]
      },
      "variants": [
        {
          "id": 263,
          "name": "Pineapple lamps",
          "sku": "11111",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 264,
          "name": "Pineapple lamps",
          "sku": "11113",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 265,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 266,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 267,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 268,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 269,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 270,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 271,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 272,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 273,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 274,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 275,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 276,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 277,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 278,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 279,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 280,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 281,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 282,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 283,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 284,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        },
        {
          "id": 285,
          "name": "Pineapple lamps",
          "sku": "",
          "price": "13.0",
          "weight": "0.0",
          "height": null,
          "width": null,
          "depth": null,
          "is_master": false,
          "slug": "pineapple-lamps-6a8d2da5-7ba7-45a1-b1e7-ae71acc9dcb0",
          "description": null,
          "track_inventory": true,
          "cost_price": null,
          "display_price": "£13.00",
          "options_text": "",
          "in_stock": false,
          "is_backorderable": true,
          "total_on_hand": 0,
          "is_destroyed": false,
          "option_values": [ ],
          "images": [ ]
        }
      ],
      "option_types": [{ "id": 1, "name": "size", "presentation": "Size", "position": 0 }],
      "product_properties": [ ],
      "classifications": [ ]
    }
  ]
}
```
I have tried the following syntax amongst others, neither work:
curl -i -X POST -H "X-Spree-Token:  XXX" http://XXX/products/$1/variants -d "variant[price]=13&variant[option_value_ids][size]=1"

curl -i -X POST -H "X-Spree-Token:  XXX" http://XXX/products/$1/variants -d "variant[price]=13&variant[option_value_ids][]=1"

**Comment by JDutil** [comment]
You may need to try upgrading to 3-0-stable this was a bug that should have been fixed. There is an example for creating variants here:
https://github.com/spree-contrib/spree_api_examples/blob/master/examples/products/variants/crud.rb

**Comment by vmatekole** [comment]
Hi, this example does not cover creating variants, with option_values. And I am indeed using Spree stable 3.0.0.
I have found the issue and the correct syntax is:

curl -i -X POST -H "X-Spree-Token: XXX" http://XXX/api/products/$1/variants -d "variant[price]=13&variant[option_value_ids]=1"

I have found looking at the requests(using Firebug or something similar) made from the admin interface is a helpful guide to discovering use cases for the API that are not documented.

Cheers.

**Comment by JDutil** [comment]
I mean try updating to the 3-0-stable branch we have recently [fixed this bug](https://github.com/spree/spree/commit/84c7bdaec9cf24fcad01816a0d1e285bafea35a6), and I don't think the fix is in Spree 3.0.0.

**Comment by vmatekole** [comment]
@JDutil  - I have rebuilt my docker image using 3-0-stable - gemfile was always set to pull from this branch. 

I have used the following combinations of curl requests and I can only assign a single option value. NOT multiple. The examples you have done use JSON - I was not aware from the documentation you could post JSON. Perhaps, there is no issue when using JSON, will check.


curl -i -X POST -H "X-Spree-Token: XXX" XXXX/api/products/$1/variants -d "variant[price]=13&variant[option_value_ids][]=1&variant[option_value_ids][]=5" - does not work 

curl -i -X POST -H "X-Spree-Token: XXX" XXXX/api/products/$1/variants -d "variant[price]=13&variant[option_value_ids]=1&variant[option_value_ids]=5" - does not work 


curl -i -X POST -H "X-Spree-Token: XXX" XXXX/api/products/$1/variants -d "variant[price]=13&variant[option_value_ids][]=1" - does not work 

curl -i -X POST -H "X-Spree-Token: XXX" XXXX/api/products/$1/variants -d "variant[price]=13&variant[option_value_ids]=1" - Works

**Comment by vmatekole** [comment]
@JDutil  - I have tried to create variants using JSON and still cannot create multiple option_value_id via an array:

## DOES NOT WORK
{
    "variant" : {
       "price" : "9.99",
       "option_value_ids": [1]
    }
}

# DOES NOT WORK
{
    "variant" : {
       "price" : "9.99",
       "option_value_ids": [1,5]
    }
}

# DOES WORK
{
    "variant" : {
       "price" : "9.99",
       "option_value_ids": 1
    }
}

**Comment by JDutil** [comment]
Will reopen until I can look into it again, but I would suggest you look into the strong parameters in the meantime as I'm not sure when I'll have time to investigate this at all as far as I can tell it's been fixed already.

**Comment by vmatekole** [comment]
Danke! Here are some scripts you can use to test against my use case

https://github.com/GROK7/spree-vend/blob/20ddc3627f11954627f81e8eadb073fb1b2186ff/scripts/create_spree_variant.sh

https://github.com/GROK7/spree-vend/blob/20ddc3627f11954627f81e8eadb073fb1b2186ff/scripts/example-variants.json

**Comment by huoxito** [comment]
@vmatekole so I was reading this thread again and apparently you haven't told what error you're getting? _does not work_ doesn't help us much, the api build has [a spec](https://github.com/spree/spree/blob/0c7213b33872bf43839f98948a8b9b801322ee37/api/spec/controllers/spree/api/variants_controller_spec.rb#L169-178) running something very similar to one of your examples, I also modified it to pass two option value ids and a price but it is still green.

could you let us know what error you getting? perhaps there's something wrong with your setup, a spec or app to reproduce the issue would be great thanks

**Comment by vmatekole** [comment]
@huoxito  apologies, please look at the response:

```json
{
  "id": 32,
  "name": "Apache Baseball Jersey",
  "sku": "",
  "price": "9.99",
  "weight": "0.0",
  "height": null,
  "width": null,
  "depth": null,
  "is_master": false,
  "slug": "apache-baseball-jersey",
  "description": "Enim deserunt architecto ullam earum nemo. Quisquam ut doloremque quia vel. Aut tenetur beatae dolorem sunt omnis ea.",
  "track_inventory": true,
  "cost_price": null,
  "display_price": "$9.99",
  "options_text": "",
  "in_stock": false,
  "is_backorderable": true,
  "total_on_hand": 0,
  "is_destroyed": false,
  "option_values": [ ],
  "images": [ ],
  "stock_items": [
    {
      "id": 32,
      "count_on_hand": 0,
      "stock_location_id": 1,
      "backorderable": true,
      "available": true,
      "stock_location_name": "default"
    }
  ]
}
```

You will notice no option_values are applied to the variant.

**Comment by JDutil** [comment]
@huoxito could you commit or pr the updated spec using multiple option values?

**Comment by huoxito** [comment]
yeah sure will push soon

On Wed, Apr 29, 2015 at 10:46 AM, Jeff Dutil <notifications@github.com>
wrote:

> @huoxito <https://github.com/huoxito> could you commit or pr the updated
> spec using multiple option values?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spree/spree/issues/6331#issuecomment-97430269>.
>



-- 
Washington L Braga Jr

Developer
Spree Commerce, Inc.

**Comment by huoxito** [comment]
@vmatekole still not able to reproduce, see this spec for example https://github.com/spree/spree/commit/3af37a8dbe3671a63a1d5890523e526407f66239 it should fail according to your report, I tried a real curl submiting the variant json on a fresh spree install and it also works fine, so I still think one of us is probably missing something, you sure you can reproduce that on a fresh spree install? any chance you could submit an example app with the issue? otherwise I guess we cant do anything about it if we cant reproduce the error

**Comment by vmatekole** [comment]
@JDutil  and @huoxito I rebuilt my docker image and redeployed to my staging server. I also did a bundle install locally, assuming that bundler does a **git pull**  from 3-0-stable branch. My assumptions were incorrect... Docker seemingly pulls from cache on the bundle install, so a **bundle update** is required in my Dockerfile, and running the same locally is also required to cause a **git pull**. 

I do apologise for the wasted time on this issue... Without wanting to sound blameless, perhaps bumping the version number for such updates would avoid confusion?
